### ISSUE TITLE
Woes with `~NULL` and rlang data masks

### ISSUE THREAD

**Issue by DavisVaughan** [HenrikBengtsson/globals#64]
Related to https://github.com/HenrikBengtsson/globals/issues/59

Unfortunately furrr has a revdep that does something like this:

``` r
library(dplyr)
library(globals)

df <- tibble(x = 1)

globalsOf(~NULL)
#> $`~`
#> .Primitive("~")
#> 
#> attr(,"class")
#> [1] "Globals" "list"   
#> attr(,"where")
#> attr(,"where")$`~`
#> <environment: base>

mutate(df, y = list(globalsOf(~NULL)))
#> Error: Problem with `mutate()` input `y`.
#> x missing value where TRUE/FALSE needed
#> ℹ Input `y` is `list(globalsOf(~NULL))`.
```

At the lowest level, this is the same problem as #59. There is a `~NULL` formula that is being subset by `x[-1]`, which causes that stats bug.

The absolute simplest reprex, without dplyr, would be this one with rlang:

``` r
library(rlang)
library(globals)

df <- data.frame(x = 1)
data_mask <- as_data_mask(df)
env <- environment()

expr <- quote(globalsOf(~NULL))

eval_tidy(expr, data = data_mask, env = env)
#> Error in if (length(ans) == 0L || as.character(ans[[1L]])[1L] == "~") {: missing value where TRUE/FALSE needed
```

`df` has been turned into a "data mask" environment and is installed as the child env of `env`. This makes it so that when `expr` is evaluated, the columns of `df` are found first, then we look up everything else in `env`.

The problem is that this data mask env always defines `~` in it:

```r
data_mask$`~`
#> function (...) 
#> {
#>     .Call(rlang_tilde_eval, sys.call(), environment(), parent.frame())
#> }
#> <environment: namespace:rlang>
```

When we run this through `globalsOf()`, eventually codetools checks to see what env `~` came from. If it came from a base R package env (like stats), it takes a specific path, otherwise it calls `codetools:::collectUsageArgs()`. That is what happened here since `~` was masked. And unfortunately you can see that codetools calls `e[-1]` in that function, which causes this issue:

```r
> codetools:::collectUsageArgs
function (e, w) 
{
    for (a in dropMissings(e[-1])) if (typeof(a) == "symbol" && 
        a == "...") {
        if (w$isLocal("...", w)) 
            w$enterLocal("variable", "...", a, w)
        else w$signal(paste(a, "may be used in an incorrect context:", 
            pasteExpr(e)), w)
    }
    else walkCode(a, w)
}
<bytecode: 0x7fc00193c1c0>
<environment: namespace:codetools>
```

**Comment by HenrikBengtsson** [comment]
I run 1st and 2nd generation revdeps, cf. https://github.com/HenrikBengtsson/globals/blob/develop/revdep/README.md.  Isn't that package on that list?  On the fresh, new system I'm testing on, package tests that attempt to open a graphics device fail so it could be that this problem has been masked by those failures.

**Comment by DavisVaughan** [comment]
baguette is the package. I'm fairly certain that it "works" with CRAN furrr, but that was because I was passing the wrong environment through to `future::getGlobalsAndPackages()`. In the development version of furrr I use the new `.env_globals` argument to pass the caller environment (`parent.frame()`) through, which I am pretty confident is the right thing to do

**Comment by DavisVaughan** [comment]
This issue seems a bit tougher to solve because its codetools itself that calls `e[-1]`. I feel like the only way to work around it is to catch and handle the `~` before it has a chance to go down that problematic codetools path

**Comment by DavisVaughan** [comment]
With just base R and globals

``` r
library(globals)
#> Warning: package 'globals' was built under R version 4.0.2

env <- new.env(parent = globalenv())

env$`~` <- function(...) {
  "OVERRIDE!"
}

x <- ~NULL

expr <- quote(findGlobals(x))

eval(expr, env)
#> Error in if (length(ans) == 0L || as.character(ans[[1L]])[1L] == "~") {: missing value where TRUE/FALSE needed
```

<sup>Created on 2020-09-18 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0.9001)</sup>

**Comment by HenrikBengtsson** [comment]
Without having looked at the details, tracebacks, etc., would you say this something that should/could be fixed in **codetools**?

**Comment by DavisVaughan** [comment]
I think ideally `stats:::[.formula` would be fixed.

If that doesn't happen, codetools could probably work around this by not using `e[-1]` in `codetools:::collectUsageArgs()`. It hasn't been updated since 2018, but maybe Luke would be responsive about getting a quick patch out?

**Comment by HenrikBengtsson** [comment]
Ok, so, the bug is indeed in **stats**:

```r
> x <- ~NULL
> x[-1]
Error in if (length(ans) == 0L || as.character(ans[[1L]])[1L] == "~") { : 
  missing value where TRUE/FALSE needed
> traceback()
2: `[.formula`(x, -1)
1: x[-1]
```

which is because `length(ans) == 1` an `ans[[1L]]` is `NULL`;

```r
> trace(stats:::`[.formula`, tracer=quote(utils::str(list(ans = ans, first = ans[[1L]]))), at=3L)
Tracing function "[.formula" in package "stats (not-exported)"
[1] "[.formula"
> x <- ~NULL
> x[-1]
Tracing `[.formula`(x, -1) step 3 
List of 2
 $ ans  : language NULL()
 $ first: NULL
Error in if (length(ans) == 0L || as.character(ans[[1L]])[1L] == "~") { : 
  missing value where TRUE/FALSE needed
```

Did you report this to BugZilla?

**Comment by HenrikBengtsson** [comment]
I've submitted [PR17935](https://bugs.r-project.org/bugzilla/show_bug.cgi?id=17935) with a straightforward patch.

**Comment by DavisVaughan** [comment]
Patch looks good! 

Do you think there is any possibility of a way globals can work around this so it works on older versions of R?

**Comment by HenrikBengtsson** [comment]
> Do you think there is any possibility of a way globals can work around this so it works on older versions of R?

I'm hoping so but haven't had time to drill down where in the global-codetools-walking maze this happens and whether or not it's possible to patch it, e.g. by updating the walker or injecting the patched version of `stats::[.formula` in there.  To be found out.

**Comment by DavisVaughan** [comment]
I think it’s the same big as this one you reported. I figure you realized that but you didn’t link to this, so I just wanted to add this into the conversation https://stat.ethz.ch/pipermail/r-devel/2020-August/079838.html

**Comment by HenrikBengtsson** [comment]
He - it's either my bad memory or too many things in parallel.  I do remember this e`[-1]` came up a few months ago but without that email archive, you could have told anything at this point :p

**Comment by HenrikBengtsson** [comment]
Update: the stats bug has been fixed in R-devel.

**Comment by HenrikBengtsson** [comment]
Fixed in the develop branch by adding a workaround for R (< 4.1.0) (commit a5505b8)

**Comment by DavisVaughan** [comment]
Wow- impressive workaround. I really appreciate you looking at this!

**Comment by HenrikBengtsson** [comment]
FWIW, the **stats** bug fix is now in R 4.0.3 patched, i.e. it'll appear in R (>= 4.0.4) whenever that's released, cf. https://github.com/wch/r-source/commit/add1000ed1008196e6f5b435ed0c8580b61cf11e
